### DISCUSSION TITLE
How to pass a color as global variable?

### DISCUSSION THREAD

**Discussion by Cicorione** [speedata/publisher#624] [Q&A] [answered]
Hi @pgundlach ,

I am trying to create some global variables to handle the colors in the `switch` cases, but it is not working as I expected, and before to open a bug a want to catch out the _pebcak_ first.

I have my global variables in the header, since is not working with my personal colors, I tried with the default ones, but the result did not change, colors aren't displayed.

My test variables (not used at the same time):

`<SetVariable variable="title-bar-color" select="blue" />`
`<SetVariable variable="title-bar-color" select="'blue'" />`

My code:
```xml
[...]
<Tr valign="middle" minheight="8mm">
    <Td>
        <Box height="8mm" width="4mm" graphic="tableedgeleft" />
    </Td>
    <Td padding-right="1mm" padding-bottom="-3pt" background-color="$title-bar-color">
        <Paragraph>
            <B>
                <Value select="upper-case(@title)" />
            </B>
        </Paragraph>
    </Td>
    <Message select="concat('BAR1: ',$title-bar-color)" />
[...]
```

The warning:
`<entry level="warn" msg="[page 3] Background: Color &#34;$title-bar-color&#34; is not defined"></entry>`

But the message is read:
`<entry level="message" msg="BAR1: blue" line="295" page="4"></entry>`

### What am I doing wrong this time? 😩

Thanks... 🙏

**Comment by pgundlach** [comment, marked as answer]
```xml
<SetVariable variable="title-bar-color" select="blue" />
<SetVariable variable="title-bar-color" select="'blue'" />
```

the first line will not work, as you select the value of the element `blue`. The second works since you select the string `"blue"`.

When you use `background-color="$title-bar-color"` you explicitly ask for the color named `$title-bar-color`  which could be defined with

```xml
<DefineColor name="$title-bar-color" ...>
```

which I assume you did not do. 

You need to use curly braces as described in [the manual](https://doc.speedata.de/publisher/en/lxpath/#_xpath_expressions_2). It jumps into XPath mode and evaluates the expression, in this case it is a variable reference (getting the contents of the variable with the name`title-bar-color`).

In the case of 

```xml
<Message select="concat('BAR1: ',$title-bar-color)" />
```

the XPath expression is evaluated by using select.

There are two cases

1) the name of the attribute is `select` or `test`: the contents of the attribute will be evaluated as an XPath expression
2) the name of the attribute is _not_ `select` or `test`: the contents of the attribute is taken verbatim. You can jump into XPath mode with curly braces.

This is why the error message is `msg="[page 3] Background: Color "$title-bar-color" is not defined"`. The command `Td` does not evaluate the attribute `background-color` as an XPath expression.

Solution:

```xml
<Td padding-right="1mm" padding-bottom="-3pt" background-color="{$title-bar-color}">
```

(untested)

**Reply by Cicorione** [reply]
Thank you Patrick, I came right here to say that I found the pebcak:

```xml
<Td padding-right="1mm" padding-bottom="-3pt" background-color="{$title-bar-color}">
```

🤦‍♂️

**Reply by Cicorione** [reply]
Hi Patrick I have the colors defined this way:
```xml
<DefineColor name="aglgray" model="cmyk" c="0" m="0" y="0" k="40" />
```

I want to keep the colors and only change the variables, I could do the other way around too but it looks for me more rationale assigning a color to a variable than creating several copies of the same color assigned to different places... 🤔

**Reply by pgundlach** [reply]
> I wan't to keep the color and only change the variables, I could do the other way around too but it looks for me more rationale assigning a color to a variable than creating several copies of the same color assigned to different places... 🤔

You can do both, just a matter of taste. No performance or memory penalty if you chose either way.
